enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -5414,18 +5572,40 @@ class AdjointGenerator
         Type *MPI_OP_Ptr_type =
             PointerType::getUnqual(Type::getInt8PtrTy(call.getContext()));
 
-        Value *count =
-            lookup(gutils->getNewFromOriginal(call.getOperand(1)), Builder2);
-        Value *datatype =
-            lookup(gutils->getNewFromOriginal(call.getOperand(2)), Builder2);
-        Value *root =
-            lookup(gutils->getNewFromOriginal(call.getOperand(3)), Builder2);
-        Value *comm =
-            lookup(gutils->getNewFromOriginal(call.getOperand(4)), Builder2);
+        Value *count = gutils->getNewFromOriginal(call.getOperand(1));
+        if (!forwardMode)
+          count = lookup(count, Builder2);
+        Value *datatype = gutils->getNewFromOriginal(call.getOperand(2));
+        if (!forwardMode)
+          datatype = lookup(datatype, Builder2);
+        Value *root = gutils->getNewFromOriginal(call.getOperand(3));
+        if (!forwardMode)
+          root = lookup(root, Builder2);
+
+        Value *comm = gutils->getNewFromOriginal(call.getOperand(4));
+        if (!forwardMode)
+          comm = lookup(comm, Builder2);
 
         Value *rank = MPI_COMM_RANK(comm, Builder2, root->getType());
```

Comment:
Move the rank/type size calls after forward mode as not needed

Reply:
done